civicrm.user.inc

```diff
@@ -11,6 +11,18 @@ use Drupal\Core\Session\AccountInterface;
  * Implements hook_user_login().
  */
 function civicrm_user_login(AccountInterface $account) {
+  $civicrm = \Drupal::service('civicrm');
+  $civicrm->initialize();
+
+  if (\Drupal::hasService('masquerade')) {
+    // In theory this is harmless whether we masquerade or not, but for now narrowing
+    // the scope, in case it causes regressions.
+    // Also, there is a isMasquerading() function, but it does not seem to be
+    // initialized at this point (it always returns false).
+    $session = \CRM_Core_Session::singleton();
+    $session->reset(FALSE);
```

Comment:
On the first (normal) login, the CiviCRM session stores the userID and ufID, but when the user masquerades and `synchronizeUser` gets called again, the old session ufID doesn't match anymore and so it does a hard reset of the session (causing a logout).

Using `$session->reset(FALSE)` does a soft reset of the CiviCRM-specific session bits, so that the call to `synchronize` then re-updates the ufID and whatever is required.

Finally, I tried to reduce the scope of this fix, in case it causes regressions (my first attempt had broken the dashlets). I would have use `\Drupal::service('masquerade')->isMasquerading()`, but for some reason it did not work and I could not figure out why.